File: scala/Euler5.scala

/*
 * 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
 * What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
 * 
*/
object Euler5 extends App {

  val gcd: (Int, Int) => Int = (a, b) => if (b == 0) a else gcd(b, a % b)
  val lcm: (Int, Int) => Int = (a, b) => a * (b / gcd(a, b))

  def solution(): Long = {
    Stream.from(1) take(20) reduceLeft(lcm(_, _))
  }

  println(solution)
}

File: python/Euler5.py
"""
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.
What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?
"""

def solution():
    return reduce(lambda x,y: lcm(x,y), range(1,20))
                
def gcd(a,b):
    if(b==0):
        return a
    else:
        return gcd(b, a%b)
    
    
def lcm(a,b):
    return (a * (b / gcd(a, b)))       

